src/soter/boringssl/soter_sign_ecdsa.c

```diff
@@ -30,52 +30,66 @@ soter_status_t soter_sign_init_ecdsa_none_pkcs8(soter_sign_ctx_t* ctx,
                                                 const void* public_key,
                                                 const size_t public_key_length)
 {
-    /* pkey_ctx init */
-    EVP_PKEY* pkey;
+    soter_status_t err = SOTER_FAIL;
+    EVP_PKEY* pkey = NULL;
+
     pkey = EVP_PKEY_new();
     if (!pkey) {
         return SOTER_NO_MEMORY;
     }
+
     if (!EVP_PKEY_set_type(pkey, EVP_PKEY_EC)) {
-        EVP_PKEY_free(pkey);
-        return SOTER_FAIL;
+        goto free_pkey;
     }
+
     ctx->pkey_ctx = EVP_PKEY_CTX_new(pkey, NULL);
```

Comment:
Functions that are not marked SOTER_API are considered to be private to Soter and they can make assumptions about arguments. Usually we check arguments for NULL in public functions and let the private ones assume that the arguments are more or less sane.

In this case soter_sign_init_ecdsa_none_pkcs8() is called by soter_sign_init() which does check that ctx is non-NULL beforehand. It’s also checked a couple of times more up the call stack, so we can be sure that it won’t be NULL in soter_sign_init_ecdsa_none_pkcs8().

SOTER_PRIVATE_API is kind of a hack that intentionally exposes _private_ API for Themis. It’s not really pretty, but some code currently calls such functions and makes necessary precautions to ensure that they are safe to call. At some point we intend to get rid of SOTER_PRIVATE_API to avoid confusion over what’s public and what’s private.

Reply:
Thanks for your explanation. Now I have knowledge how this visibility works. This kind of bug is not connected with PR. But you have such functions in SOTER_API. Ex. is soter_rsa_key_pair_gen_init(). This is notice to be cerfull with this.

Reply:
Mb it will be good to make some rule how to work with this.

Reply:
Ok, with soter_rsa_key_pair_gen_init() is bad example (becouse it should be not in SOTER_API).   But if we look at typtcal usage any public funtion it would be checked twice ore even more.

Reply:
Well, C does not really provide any amenities for null checks. Theoretically, we could use `__attribute__((nonnull))` or something like that, but that’s not really portable and too much of a bother. Generally, the rules are that public functions (marked with SOTER_API or THEMIS_API) have to do ~~paranoid~~ thorough parameter checking while internal functions may expect the parameters to have reasonable values. And generally, it does not really hurt if internal functions have duplicated checks. Better safe than segfault.